src/utils/elf.rs

```diff
@@ -28,3 +30,50 @@ pub fn extract_btf_from_file<P: AsRef<Path>>(path: P) -> Result<Vec<u8>> {
 
     Ok(btf.to_vec())
 }
+
+/// Given a directory containing a 'vmlinux' ELF file in its root and optional
+/// '*.ko' ELF modules in the root or any sub-directory (this maps well to a
+/// Linux build directory or /usr/lib/modules/), initialize a BtfCollection
+/// extracting BTF data from the .BTF section of those files.
+pub fn collection_from_kernel_dir<P: AsRef<Path>>(path: P) -> Result<BtfCollection> {
+    let path = path.as_ref();
+    if !path.is_dir() {
+        bail!(
+            "Can't initialize a BTF collection from {}: not a directory",
+            path.display()
+        );
+    }
+
+    // Find the base BTF file and initialize the collection.
+    let vmlinux = path.join("vmlinux");
+    let mut collection = BtfCollection::from_bytes("vmlinux", &extract_btf_from_file(vmlinux)?)?;
+
+    // Traverse the directory looking for modules.
+    fn visit_dir<P: AsRef<Path>>(dir: P, collection: &mut BtfCollection) -> Result<()> {
+        for entry in fs::read_dir(dir)? {
+            let path = entry?.path();
+            let filename = path
+                .file_name()
+                .ok_or_else(|| anyhow!("Could not get module file name"))?
+                .to_str()
+                .ok_or_else(|| anyhow!("Could not convert module name to str"))?;
+
+            if path.is_dir() {
+                visit_dir(path, collection)?;
+            } else if filename.ends_with(".ko") {
+                collection.add_split_btf_from_bytes(
+                    match filename.split_once('.') {
+                        Some((name, _)) => name,
+                        // Should not happen as we already filtered on extensions.
+                        None => bail!("Invalid module file name"),
+                    },
+                    &extract_btf_from_file(&path)?,
```

Comment:
What if one of the modules do not have btf information? Is that a reasonable scenario? should we fail in that case or just ignore/warn?

Reply:
Yep, I guess failing is better than having partial btf info.